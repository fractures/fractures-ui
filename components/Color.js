import { Component } from "react"
import cc from "classcat"

const defaultProps = {
	hex: "",
	name: ""
}

class Color extends Component {
	render() {
		const colorClasses = cc({
			"grow-1 p-2 nowrap": true,
			white: this.props.name === 4
		})

		return (
			<small className={ colorClasses } style={ { backgroundColor: this.props.hex } }>
				<b className="nowrap block">{this.props.name}</b>
				<span className="opacity-50">{this.props.hex}</span>
			</small>
		)
	}
}

export default Color
