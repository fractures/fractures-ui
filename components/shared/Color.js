import React, { Component } from "react"

class Color extends Component {
	render() {
		return (
			<small className="fr-color grow-1 p-2 nowrap white" style={ { backgroundColor: this.props.hex } }>
				{this.props.hex}
				<style jsx>{`
					.fr-color {
						text-shadow: 0 1px 0 rgba(0, 0, 0, 0.2);
					}
				`}</style>
			</small>
		)
	}
}

export default Color
