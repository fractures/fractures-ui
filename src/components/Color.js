import React, { Component } from "react"

const Color = () => (
	<small className="fr-color grow-1 p-2 nowrap white" style={ { backgroundColor: props.hex } }>
		<b>{props.name}</b>
		<br />
		<small>{props.hex}</small>
		<style jsx>{`
			.fr-color {
				text-shadow: 0 1px 0 rgba(0, 0, 0, 0.2);
			}
		`}</style>
	</small>
)

export default Color
