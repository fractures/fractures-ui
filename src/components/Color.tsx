import * as wcag from 'wcag-contrast'
import chroma from 'chroma-js'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

const propTypes = {
	background: PropTypes.string.isRequired,
	hex: PropTypes.string.isRequired,
	isVerbose: PropTypes.bool,
	name: PropTypes.string.isRequired,
	variant: PropTypes.number.isRequired
}

const defaultProps = {
	name: '',
	isVerbose: false,
	variant: 0
}

// Get the rounded color contrast value
const colorValue = (color, background = 'white') => {
	const normalizedColor = chroma(color).hex()
	const normalizedBackgroundColor = chroma(background).hex()

	const wcagScore = wcag.hex(normalizedColor, normalizedBackgroundColor)
	const contrastNumber = Number(wcagScore).toFixed(2)

	return contrastNumber
}

// Get color contrast score from contrast score value
const colorScore = contrastValue => {
	let score

	if(contrastValue > 7) score = `AAA`
	else if(contrastValue >= 4.51) score = `AA`
	else if(contrastValue < 4.51 && contrastValue > 3) score = `AA L`
	else score = `F`

	return score
}

// Normalize input color
// const normalizeColor = color => {
// 	const normalizedColor = chroma(color).rgb()
//
// 	return `R: ${ normalizedColor[0] } G: ${ normalizedColor[1] } B: ${ normalizedColor[2] }`
// }

const NakedColor = props => {
	const calcValue = colorValue(props.hex, props.background)
	const calcScore = colorScore(calcValue)
	const hasFailingContrast = Boolean(calcScore === 'F')
	const color = hasFailingContrast ? 'currentColor' : props.hex

	return (
		<div className={ props.className }>
			<div className="hl-color__box" style={ { backgroundColor: props.hex } } />
			{props.isVerbose && (
				<div className="hl-color__a11y" style={ { color: color } }>
					<small>
						<b>{calcScore}</b>
					</small>
					<small>{calcValue}</small>
				</div>
			)}
			<div className="hl-color__meta" style={ { color: color } }>
				<p>
					<b>
						{props.name} {props.variant}
					</b>
				</p>
				<small>
					{props.hex}
					{/* {normalizeColor(props.hex)} */}
				</small>
			</div>
		</div>
	)
}

const Color = styled(NakedColor)`
	display: flex;

	color: var(--color-gray-900);

	.hl-color__box {
		height: 3rem;
		min-width: 3rem;
		width: 3rem;
	}

	.hl-color__a11y {
		display: flex;
		flex-direction: column;
		height: 3rem;
		justify-content: center;
		min-width: 3rem;
		padding: 0 0.5rem;
		width: 3rem;
	}

	.hl-color__meta {
		display: flex;
		flex-direction: column;
		justify-content: center;
		margin-left: 1rem;
		margin-right: 1rem;
		width: 6rem;
	}
`

Color.propTypes = propTypes
Color.defaultProps = defaultProps

export default Color
