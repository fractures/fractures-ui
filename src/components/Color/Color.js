import * as wcag from "wcag-contrast"
import chroma from "chroma-js"
import Label from "../Label/Label"
import Small from "../Type/Small"
import Code from "../Type/Code"
import P from "../Type/P"
import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"

const propTypes = {
	background: PropTypes.string.isRequired,
	hex: PropTypes.string.isRequired,
	isVerbose: PropTypes.bool,
	name: PropTypes.string.isRequired
}

const defaultProps = {
	isVerbose: false
}

// Get the rounded color contrast value
const colorValue = (color, background = "white") => {
	const normalizedColor = chroma(color).hex()
	const normalizedBackgroundColor = chroma(background).hex()

	const wcagScore = wcag.hex(normalizedColor, normalizedBackgroundColor)
	const contrastNumber = Number(wcagScore).toFixed(2)

	return contrastNumber
}

// Get color contrast score from contrast score value
const colorScore = contrastValue => {
	let score

	if(contrastValue > 7) score = `AAA`
	else if(contrastValue >= 4.51) score = `AA`
	else if(contrastValue < 4.51 && contrastValue > 3) score = `AA Large`
	else score = `F`

	return score
}

// Normalize input color
const normalizeColor = color => {
	const normalizedColor = chroma(color).rgb()

	return `R: ${ normalizedColor[0] } G: ${ normalizedColor[1] } B: ${ normalizedColor[2] }`
}

const NakedColor = props => {
	const calcValue = colorValue(props.hex, props.background)
	const calcScore = colorScore(calcValue)
	const isInvert = calcScore === "F"

	return (
		<div className={ props.className }>
			<div className="fr-color__box" style={ { backgroundColor: props.hex } } />
			<div className="fr-color__meta" style={ { color: props.hex } }>
				<P>
					<b>{calcScore}</b>
				</P>
				<Small>{calcValue}</Small>
			</div>
			<aside>
				<header>
					<P>{props.name}</P>
					{isInvert && <Label label="Low contrast" />}
				</header>
				<div>
					<Code isSmall={ true }>
						{props.hex} &mdash;{` `}
						{normalizeColor(props.hex)}
					</Code>
				</div>
			</aside>
		</div>
	)
}

const Color = styled(NakedColor)`
	display: flex;
	flex-wrap: wrap;

	color: var(--fr-900);

	font-size: 0.9125rem;

	.fr-color__box {
		height: 3rem;
		min-width: 3rem;
		width: 3rem;

		border-radius: 100%;
	}

	.fr-color__meta {
		display: flex;
		flex-direction: column;
		justify-content: center;
		margin-left: 1rem;
		margin-right: 1rem;
		width: 6rem;
	}

	aside div {
		align-items: center;
		display: flex;
	}

	aside div small {
		margin-right: 0.5rem;
	}

	aside header {
		display: flex;
		align-items: center;
	}

	aside header p {
		margin-right: 0.5rem;
	}

	b,
	small {
		display: block;
	}
`

NakedColor.propTypes = propTypes
NakedColor.defaultProps = defaultProps

Color.propTypes = propTypes
Color.defaultProps = defaultProps

export default Color
