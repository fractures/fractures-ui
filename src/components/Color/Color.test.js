import Color from "./Color.js"

describe("Color", () => {
	const componentRender = <Color background="#fff" hex="#eee" name="Light" />
	const component = renderer.create(componentRender)
	const tree = component.toJSON()

	it("Renders correctly", () => {
		expect(tree).toMatchSnapshot()
	})
})
