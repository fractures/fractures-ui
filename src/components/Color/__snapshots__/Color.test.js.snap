// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color Renders correctly 1`] = `
<div
  className="sc-ifAKCX ieNlBV"
>
  <div
    className="fr-color__box"
    style={
      Object {
        "backgroundColor": "#eee",
      }
    }
  />
  <div
    className="fr-color__meta"
    style={
      Object {
        "color": "#eee",
      }
    }
  >
    <p
      className="sc-bxivhb cfzrAM"
    >
      <b>
        F
      </b>
    </p>
    <small
      className="sc-bwzfXH hFESPj"
    >
      1.16
    </small>
  </div>
  <aside>
    <header>
      <p
        className="sc-bxivhb cfzrAM"
      >
        Light
      </p>
      <div
        className="sc-bdVaJa hBHdgd fr-label--default"
      >
        <span>
          Low contrast
        </span>
      </div>
    </header>
    <div>
      <code
        className="sc-htpNat cjHBxU"
      >
        #eee
         â€”
         
        R: 238 G: 238 B: 238
      </code>
    </div>
  </aside>
</div>
`;
